### ISSUE TITLE
Movemove performance

### ISSUE THREAD

**Issue by corradio** [tmrowco/electricitymap-contrib#2304]
I've noticed a performance degradation especially when I move the mouse around the map as there seems to be some lag that affects the tooltip update as its associated animations.
Has anyone experienced that? Maybe we're sending a lot of redux actions at each move move?

cc @fbarl

**Comment by fbarl** [comment]
I haven't noticed this personally on my Chrome, but it's possible some performance leaks were introduced in https://github.com/tmrowco/electricitymap-contrib/pull/2276.

I'll assign myself to investigate.

**Comment by fbarl** [comment]
By the way, if Redux is the bottleneck (which is quite possible), then addressing https://github.com/tmrowco/electricitymap-contrib/issues/2288 would also fix this.

**Comment by corradio** [comment]
I've noticed that the SHOW_TOOLTIP event is dispatched for each movemove on the countrytable.
It would suffice to have it dispatched on mouveover

**Comment by fbarl** [comment]
> I've noticed that the SHOW_TOOLTIP event is dispatched for each movemove on the countrytable. It would suffice to have it dispatched on mouveover.

Yes, we could do that indeed as a quickfix if we're ok with the tooltip not following the mouse cursor when hovering within a country - if that sounds like an ok compromise to you @corradio, I can put up a quick PR to see how it feels.

I've realized that it'd be very difficult to embed map tooltips deeper in DOM (and thus bypass Redux) as a part of #2288 before moving `Map` into a React component (see #2309), so maybe the change you're suggesting is indeed a way to go for now.

**Comment by corradio** [comment]
It's not urgent so I would instead try to make sure the tooltips are embedded deeper into the DOM, potentially allowing for local state.
Another things is that maybe we should decouple the update of the tooltip content, with the positioning of it? Maybe that helps.
In any case, let's indeed try to bypass redux for the tooltips.

**Comment by fbarl** [comment]
@corradio I wonder if you are still experiencing this now that #2434 has been merged?

**Comment by corradio** [comment]
Unfortunately it's not. The test I'm making is that I'm moving my mouse quickly inside the same zone, and ideally the tooltip should follow closely the mouse position. However in this case there's some lag where the tooltip position lags behind the mouse position.

**Comment by fbarl** [comment]
Ok I tried to dig a bit deeper into this and it seems that the real bottleneck are the layers, specifically the `ExchangeLayer` even when it's not being rerendered - my suspicion is that it's either that `ReactMapGL` depends on it internally for mouse move calculations or the arrow animations which now interfere with the map render cycle more (if less arrows are drawn, the lag seems to be lower).

However, there is some low hanging fruit that still improves the performance a bit so I'd still make a PR with the following changes and see where it gets us:

* Wrap `ZoneMap` with `React.memo` to cut down on the number of times it rerenders
* Use better rerender criteria for the layers (combine `useMemo` with `MapLayer` logic)
* Don't render arrows that are supposed to be hidden instead of setting `display: none` on them

**Comment by fbarl** [comment]
> my suspicion is that it's either that ReactMapGL depends on it internally for mouse move calculations or the arrow animations which now interfere with the map render cycle more

Ok, the bottleneck of this issue definitely seems to be the fact that the arrows are animated, which makes it very related to https://github.com/tmrowco/electricitymap-contrib/issues/1590. I tried replacing just the [arrows image src](https://github.com/tmrowco/electricitymap-contrib/blob/3c0168f2b963109d2657311c516e9524ea835074/web/src/components/layers/exchangelayer.js#L57) `images/${prefix}arrow-${intensity}-animated-${speed}.gif` with different values and results are quite interesting:

* Using a static image e.g. `images/brightmode.svg` yields a much better mouse move performance
* Using a fixed fast animated arrow `images/arrow-400-animated-2.gif` doesn't really change the performance but (!) using the slowly animated fixed arrow everywhere `images/arrow-400-animated-0.gif` is almost as good on my machine as using static images

It seems like the latter point might actually be due to the higher framerate of fast arrows as they rotate through the same set of frames but within a shorter time - if we could just fix the framerate of all arrows to what it currently is for the slow ones in `generate-arrows.js`, we would probably already achieve a much better performance! :partying_face:

**Comment by corradio** [comment]
Wow that's crazy. Is that also the case with CSS?

**Comment by fbarl** [comment]
What I wrote above seems to be true on my Chrome, but it only partly applies to my Firefox - I think a further performance analysis is probably required to determine which optimizations would work best generally, so I'm gonna deassign myself for now as I'm probably not gonna be the one looking into it soon.

> Is that also the case with CSS?

CSS seemed to have improved the performance on my Chrome but made it worse on my Firefox, or was it the other way around? :thinking: I can't remember exactly...

**Comment by corradio** [comment]
A few notes here:
- Removing the `position` attribute from `<Tooltip>` rendered in `<MapCountryTooltip>` removes the FPS drop (so it doesn't come from `<MapCountryTooltip>` itself. It also doesn't come from the other attributes passed to `<Tooltip>`.
- Having a single string as children of `<Tooltip>` still causes the FPS to drop (therefore, it doesn't come from the children)
- Passing a fixed `{x:500,y:200}` position doesn't cause FPS drops (so it doesn't come from the calculation inside `<Tooltip>`).
- Commenting/uncommenting the `style.transform` assignment in `<Tooltip>` enables/disables FPS drop.
- Removing arrow layer still causes a significant FPS drop (down to ~21 fps on my machine)

**Comment by Kongkille** [comment]
@corradio are you able to record a video of the lag? I can't seem to reproduce this on my V8-powered supercomputer
